loader/src/reference_binder.py

```diff
@@ -169,43 +150,37 @@ def load_cached_references(self, target_ref: str) -> DefaultDict[tuple, list]:
             pending_refs[tuple(source_ref)].append(ref)
         return pending_refs
 
-    @staticmethod
-    def identifier_to_key(resource_type, identifier):
-        value = identifier.get("value")
-        system = identifier.get("system")
-        return f"{resource_type}:{value}:{system}"
-
-    @staticmethod
-    def partial_identifier(identifier):
-        return {
-            "identifier.value": identifier.get("value"),
-            "identifier.system": identifier.get("system")
-        }
-
-    def build_find_predicate(self, resource_ids, reference_path, identifier, is_array):
-        """Finds all resources with unresolved references to a given identifier"""
-        res = {
-            "id": {
-                "$in": resource_ids
-            }
-        }
+    def unresolved_resources_filter(self, reference_path, identifier, refs, is_array):
+        query = {"id": {"$in": refs}}
         if is_array:
-            res[reference_path] = {
+            query[reference_path] = {
                 "$elemMatch": self.partial_identifier(identifier)
             }
-        else:
-            res[f"{reference_path}.identifier.value"] = identifier.get("value")
-            res[f"{reference_path}.identifier.system"] = identifier.get("system")
-        return res
+        return query
 
     @staticmethod
-    def build_update_predicate(reference_path, fhir_object, is_array):
+    def reference_update(reference_path, fhir_object, is_array):
         if is_array:
-            target_path = f"{reference_path}.$[ref].reference"
+            target_path = f"{reference_path}.$.reference"
         else:
             target_path = f"{reference_path}.reference"
         return {
             "$set": {
                 target_path: f"{fhir_object['resourceType']}/{fhir_object['id']}"
             }
         }
+
+    @staticmethod
+    def identifier_to_key(resource_type, identifier):
+        value = identifier.get("value")
+        system = identifier.get("system")
+        # Default separators include whitespaces
+        _id = json.dumps((value, system), separators=(',', ':'))
+        return f"{resource_type}:{_id}"
+
+    @staticmethod
+    def partial_identifier(identifier):
+        return {
```

Comment:
Do you think we should raise an error if value or system is absent?

Reply:
or at least log a warning ?